iroha/src/queue.rs

```diff
@@ -1,216 +1,162 @@
-        }
-    }
-}
-
-#[async_trait::async_trait]
-impl<W: WorldTrait> Handler<GetPendingTransactions> for Queue<W> {
-    type Result = PendingTransactions;
-    async fn handle(
-        &mut self,
-        GetPendingTransactions: GetPendingTransactions,
-    ) -> PendingTransactions {
-        self.pending_txs()
-    }
-}
 
-impl<W: WorldTrait> Queue<W> {
-    /// Get cloned txs that are currently in a queue.
-    pub fn pending_txs(&self) -> PendingTransactions {
-        self.pending_tx_by_hash
-            .values()
-            .cloned()
+    /// Returns all pending transactions paginated
+    pub fn waiting(&'_ self) -> impl Iterator<Item = Transaction> + '_ {
+        self.txs
+            .iter()
+            .map(|e| e.value().clone())
             .map(VersionedAcceptedTransaction::into_inner_v1)
             .map(Transaction::from)
-            .collect()
     }
 
-    /// Puts new tx into queue.
+    /// Pushes transaction into queue
     /// # Errors
-    /// Returns error if queue is full.
-    pub fn push_pending_tx(&mut self, tx: VersionedAcceptedTransaction) -> Result<()> {
-        if let Some(transaction) = self.pending_tx_by_hash.get_mut(&tx.hash()) {
-            let mut signatures: BTreeSet<_> = transaction
-                .as_inner_v1()
-                .signatures
-                .iter()
-                .cloned()
-                .collect();
-            let mut new_signatures: BTreeSet<_> =
-                tx.into_inner_v1().signatures.into_iter().collect();
-            signatures.append(&mut new_signatures);
-            transaction.as_mut_inner_v1().signatures = signatures.into_iter().collect();
-            Ok(())
-        } else if self.pending_tx_hash_queue.len() < self.max_txs_in_queue {
-            self.pending_tx_hash_queue.push_back(tx.hash());
-            let _result = self.pending_tx_by_hash.insert(tx.hash(), tx);
-            Ok(())
-        } else {
-            Err(error!("The queue is full."))
+    /// Returns transaction if queue is full
+    #[allow(
+        clippy::unwrap_in_result,
+        clippy::unwrap_used,
+        clippy::missing_panics_doc
+    )]
+    pub fn push(
+        &self,
+        tx: VersionedAcceptedTransaction,
+    ) -> Result<(), VersionedAcceptedTransaction> {
+        if self.txs.len() >= self.max_txs {
+            return Err(tx);
+        }
+
+        let hash = tx.hash();
+        let entry = match self.txs.entry(hash) {
+            Entry::Occupied(mut old_tx) => {
+                // MST case
+                old_tx
+                    .get_mut()
+                    .as_mut_inner_v1()
+                    .signatures
+                    .append(&mut tx.into_inner_v1().signatures);
+                return Ok(());
+            }
+            Entry::Vacant(entry) => entry,
+        };
+
+        entry.insert(tx);
+        match self.queue.push(hash) {
+            Ok(()) => Ok(()),
+            Err(hash) => Err(self.txs.remove(&hash).unwrap().1),
```

Comment:
```suggestion
            Ok(()) => Ok(()),
            Err(hash) => Err(self.txs.remove(&hash).expect("Hash is guaranteed to be there").1),
```

Reply:
The point is that insertion of that thing happens just before match. I think that there is no reason to add context

Reply:
If this suddenly fails - it will be easier for us to remember and understand if we have it like that.

Reply:
As a rule of thumb it is in general always better to use `expect` instead of `unwrap` if it is not a prototype I think.